nomad/state/state_store.go

```diff
@@ -2122,7 +2122,9 @@ func (s *StateStore) CSIVolumeByID(ws memdb.WatchSet, namespace, id string) (*st
 	if err != nil {
 		return nil, fmt.Errorf("volume lookup failed: %s %v", id, err)
 	}
-	ws.Add(watchCh)
+	if ws != nil {
+		ws.Add(watchCh)
+	}
```

Comment:
I like that this is explicit, I always get nervous when handling watchset. Want to note that this is not required - many other cases don't do a nil check, and apparently it's handled in Add: https://github.com/hashicorp/go-memdb/blob/v1.3.0/watch.go#L16-L20

Interestingly, this commit adds a new `ws.Add` without doing a nil check ;-).

Reply:
Ah, and it's a documented part of the `ws.Add` method. Maybe we should just remove these extra checks then?

Reply:
Done in 88a7e71

Reply:
So that's interesting... removing those caused NPE errors at those lines in the test runs for `jobVersionByID`!

Reply:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1a928e1]

goroutine 54 [running]:
github.com/hashicorp/nomad/nomad/state.(*StateStore).jobVersionByID(0xc00062b470, 0xc00057e180, 0x0, 0xc00054b630, 0x7, 0xc000551380, 0x28, 0x0, 0x0, 0x0, ...)
	github.com/hashicorp/nomad/nomad/state/state_store.go:1842 +0x161
github.com/hashicorp/nomad/nomad/state.(*StateStore).upsertJobVersion(0xc00062b470, 0x7, 0xc00064e900, 0xc00057e180, 0x0, 0x0)
	github.com/hashicorp/nomad/nomad/state/state_store.go:1753 +0x249
github.com/hashicorp/nomad/nomad/state.(*StateStore).upsertJobImpl(0xc00062b470, 0x7, 0xc00064e6c0, 0x35de500, 0xc00057e180, 0x0, 0x1)
	github.com/hashicorp/nomad/nomad/state/state_store.go:1523 +0x42c
github.com/hashicorp/nomad/nomad/state.(*StateStore).UpsertJob(0xc00062b470, 0x4, 0x7, 0xc00064e6c0, 0x0, 0x0)
	github.com/hashicorp/nomad/nomad/state/state_store.go:1449 +0xa8
github.com/hashicorp/nomad/nomad.(*nomadFSM).applyUpsertJob(0xc0005f5030, 0x4, 0xc000391801, 0x6b4, 0xbb3, 0x7, 0x0, 0x0)
	github.com/hashicorp/nomad/nomad/fsm.go:507 +0x205
github.com/hashicorp/nomad/nomad.(*nomadFSM).Apply(0xc0005f5030, 0xc000241a90, 0x51dd880, 0xbfe7b879c156277c)
	github.com/hashicorp/nomad/nomad/fsm.go:219 +0x34e
github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc0006a43e0)
	github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:90 +0x2c2
github.com/hashicorp/raft.(*Raft).runFSM.func2(0xc000692e00, 0x1, 0x40)
	github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:113 +0x75
github.com/hashicorp/raft.(*Raft).runFSM(0xc0005a8000)
	github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/fsm.go:219 +0x3c4
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc0005a8000, 0xc0004c4100)
	github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:146 +0x55
created by github.com/hashicorp/raft.(*raftState).goFunc
	github.com/hashicorp/raft@v1.1.3-0.20200211192230-365023de17e6/state.go:144 +0x66
    jobs_test.go:2176: err: Put "http://127.0.0.1:31070/v1/jobs": EOF
```

Reply:
Oh, I think we might have found a bug in go-memdb. That NPE is because that nil check doesn't do a lick of good... https://play.golang.org/p/a3u09SNTUFX

Reply:
Wow - TIL.  Thanks for digging in!

Reply:
I'm going to revert the change in this PR and open a new PR adding the check to other watches.

Reply:
Hm, I must be missing something here, because testing with the actual go-memdb library shows otherwise...

```
func TestNilWatch(t *testing.T) {

	test := func(ws WatchSet) {
		watchCh := make(chan struct{})
		ws.Add(watchCh)
	}
	test(nil)
	test(NewWatchSet())
}
```

```
$ go test . -v -run TestNilWatch -count=1
=== RUN   TestNilWatch
--- PASS: TestNilWatch (0.00s)
PASS
ok      github.com/hashicorp/go-memdb   0.025s
```

Reply:
Oh, it's because `jobVersionByID`'s signature is:

```go
(s *StateStore) jobVersionByID(txn *txn, ws *memdb.WatchSet, namespace, id string) ([]*structs.Job, error)
```

I don't think we're supposed to be taking that pointer?

Reply:
Interesting `jobVersionByID` takes a `*memdb.WatchSet` unlike the other methods in this file.  It was introduced as such since the beginning in https://github.com/hashicorp/nomad/pull/2566/files#diff-53c7d524897a192ece819816a9dcf006dee46747fbb853e7f638601ea556a72cR561-R572 - but without explanation...

Reply:
Given that it's the only place in the code base where we do that, including Consul and Vault, I suspect it's just a long-lived typo.

Reply:
Oops - I missed the last comment when I posted mine - I agree, it shouldn't be a pointer here.